README.md
# ChatResearch

[![pypi](https://img.shields.io/pypi/v/chat-research.svg)][pypi status]
[![python version](https://img.shields.io/pypi/pyversions/chat-research)][pypi status]
[![license](https://img.shields.io/pypi/l/chat-research)][license]
[![precommit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)][precommit]
[![black](https://img.shields.io/badge/code%20style-black-000000.svg)][black]
[![Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/charliermarsh/ruff/main/assets/badge/v1.json)](https://github.com/charliermarsh/ruff)

ChatResearch is a tool that uses OpenAI's GPT-3 to accelerate your research.
It provides several features such as generating summary papers, fetching and summarizing papers from arxiv and bioarxiv, generating responses for review comments, and more.

## :question: Why ChatResearch

There have been countless projects and research initiatives in the field of ChatGPT,
but none have fulfilled my particular needs.
Consequently, I have made the decision to develop my own project.
I am committed to continuously refining and improving this project.
Your support and ratings are greatly appreciated, so please do not hesitate to leave a star.
I also welcome any suggestions, proposals, and pull requests.

## :rocket: News

- Multi thread and Asynchronous support has been added.
- GIFs have been added for documentation.
- Config can now be used to set key.
- Output for Latex and PDF is now supported.

## :bookmark: TODO

- [x] Multi thread and Asynchronous support
- [x] Add GIF for document
- [x] Use config to set key
- [x] Support output for Latex and PDF
- [ ] Fine-tune prompt
- [ ] Generate Image
- [ ] Add RSS support for multiple journal
- [ ] Revise help message
- [ ] add classic paper as examples

## :star2: Features

- [Chat Config](#chat-config): Generate `chatre.toml` in current working directory or set environment variable `OPENAI_API_KEY`.
- [Chat Paper](#chat-paper): Fetch or summary paper from local or arxiv with specified query, research fields, and language.
- [Chat Biorxiv](#chat-biorxiv): Fetch and summary paper from bioarxiv with specified category, filter keys, and language.
- [Chat Reviewer](#chat-reviewer): Generate summary paper with specified research fields and language.
- [Chat Response](#chat-response): Generate response for review comment with specified language.
- [Markdown](https://github.com/cauliyang/ChatResearch/blob/main/docs/examples/example2.md), [PDF](https://github.com/cauliyang/ChatResearch/blob/main/docs/examples/example1.pdf), and [Latex](https://github.com/cauliyang/ChatResearch/blob/main/docs/examples/example5.tex)

## :gear: Installation

```console
pip install chat-research
```

## :face_with_monocle: Usage

<img src="https://github.com/cauliyang/ChatResearch/blob/main//docs/tutorial/help.gif" width="800" height="400">

### Chat Config

<img src="https://github.com/cauliyang/ChatResearch/blob/main//docs/tutorial/config.gif" width="800" height="400">

```console
❯ chatre config
```

It will generate `chatre.toml` in current working directory. Otherwise,
setting environment variable `OPENAI_API_KEY` is another way to config API KEY.
Also, `chatre config set OPENAI_API_KEY=sk-key` will create configuration file
in current file directory and set your key as `sk-key`.

```toml
[OpenAI]
OPENAI_API_KEYS = [ "sk-key1", "sk-key2",]

[Gitee]
api = "your_gitee_api"
owner = "your_gitee_name"
repo = "your_repo_name"
path = "files_name_in_your_repo"
```

### Chat Paper

<img src="https://github.com/cauliyang/ChatResearch/blob/main//docs/tutorial/paper.gif" width="800" height="400">

### Chat Biorxiv

<img src="https://github.com/cauliyang/ChatResearch/blob/main//docs/tutorial/biorxiv.gif" width="800" height="400">

### Chat Response

<img src="https://github.com/cauliyang/ChatResearch/blob/main//docs/tutorial/response.gif" width="800" height="400">

### Chat Reviewer

```console
> chatre reviewer -h
usage: chatre reviewer [-h] --paper-path  [--file-format] [--review-format] [--research-fields] [--language]

optional arguments:
  -h, --help          show this help message and exit
  --paper-path        path of papers
  --file-format       output file format (default: txt)
  --review-format     review format
  --research-fields   the research fields of paper (default: computer science, artificial intelligence and reinforcement learning)
  --language          output language, en or zh (default: en)
```

## :facepunch: Develop Toolkit

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.

## Contributors

- Yangyang Li

![Alt](https://repobeats.axiom.co/api/embed/05714e0999ba02041b2be2307b7923d497c0b5c3.svg "Repobeats analytics image")

## :clap: Acknowledgement

- [ChatPaper](https://github.com/kaixindelele/ChatPaper/blob/main/README.md)
- [Arxiv](https://github.com/lukasschwab/arxiv.py)

<!-- links -->

[black]: https://github.com/psf/black
[license]: https://opensource.org/licenses/mit
[precommit]: https://github.com/pre-commit/pre-commit
[pypi status]: https://pypi.org/project/chat-research


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.1.12] - 2023-04-10

### Bug Fixes

- 🐛 fix(chat_async_arxiv.py): fix import paths
- 🐛 fix(paper_with_image.py): fix get_title method to use self.pdf instead of doc variable
- 🐛 fix(areader.py): rename find_title to update_title
- 🐛 fix(paper_with_image.py): raise exception if no image found in pdf
- 🐛 fix(areader.py): set paper title only if it is empty and a new title is found

### Co-authored-by

- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Yangliz5 <yangyang.li@northwestern.edu>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
- Dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

### Documentation

- Add example file

### Features

- 🔧 chore(cliff.toml): update commit_parsers regex to match any commit message starting with feat, fix, doc, perf, refactor, style, test, chore, and ending with any character. Update filter_unconventional to false and split_commits to true.
- 🚀 feat(.opencommitignore): add file to ignore certain file types in commits
- ✨ feat(paper_with_image.py): add fetch_title method to fetch title from pdf metadata
- Add export to PDF feature using custom template

### Miscellaneous Tasks

- 🔧 chore(Makefile): replace aic with oc in commit script
- 🔥 chore(paper_with_image.py): remove unused comments and variables
- 🔥 chore(.gitignore): add 'download' to .gitignore
- 🔖 chore(pyproject.toml): bump up version from 0.1.10 to 0.1.11

### Refactor

- Add short parameter
- 🔥 refactor(paper.py): remove unused comment
- 🔥 refactor(paper_with_image.py): remove unused comment and whitespace
- 🔨 refactor(paper_with_image.py): remove unused code and add logging statements
- 🔥 refactor(areader.py): remove unnecessary blank line
- 🔥 refactor(chat_async_paper.py): remove unnecessary comments and blank lines
- 🔨 refactor(areader.py): simplify conditional statement for updating paper title

### Signed-off-by

- Dependabot[bot] <support@github.com>

### Testing

- Bump pytest from 7.2.2 to 7.3.0 (#17)

### Build

- Bump poetry from 1.4.1 to 1.4.2 in /.github/workflows
- Bump openai from 0.27.2 to 0.27.4 (#14)
- Bump ruff from 0.0.260 to 0.0.261 (#13)
- Bump pypa/gh-action-pypi-publish from 1.5.1 to 1.8.5 (#11)
- Bump poetry from 1.4.1 to 1.4.2 in /.github/workflows (#15)
- Bump loguru from 0.6.0 to 0.7.0 (#16)

## [0.1.10] - 2023-04-05

### Features

- Add fetch_title method to fetch title from pdf metadata

### Miscellaneous Tasks

- 🔖 chore(pyproject.toml): bump up version to 0.1.10

### Refactor

- Comment out unused code

## [0.1.9] - 2023-04-05

### Documentation

- Add docs
- Add Makefile with clean, format, sync, test, and commit commands, and update README with classic paper examples in TODO list; remove unused Docker-related files and images; move tutorial images to docs folder
- Add docs

## [0.1.8] - 2023-04-05

### Bug Fixes

- 🐛 fix(AsyncBaseReader): change logger level from info to trace in report_token_usage method
- 🐛 fix(async_arxiv.py): change logger level from trace to info in Result class's download_pdf method

### Features

- 📝 docs(README.md): update README.md with new features and GIF tutorials

### Refactor

- Refactor

## [0.1.7] - 2023-04-05

### Bug Fixes

- 🔨 chore(**main**.py): fix typo in chat_arxiv_command variable name
- 🐛 fix(chat_research): rename cli function to \_cli to avoid name collision
- 🐛 fix(chat_async_paper.py): rename subcommand name from asyncpaper to paper

### Features

- Add async support
- ✨ feat(chat_research): add try-except block to handle KeyboardInterrupt

### Miscellaneous Tasks

- 🔖 chore(pyproject.toml): bump up version to 0.1.6

### Refactor

- Add aiohttp and fake-useragent to Pyproject.toml, and refactor chat_biorxiv.py and async_biorxiv.py to use async/await and aiohttp for downloading PDFs in an asynchronous way
- 🔥 refactor(**main**.py): remove chat_paper module and its references
- Refactor

## [0.1.6] - 2023-04-04

### Bug Fixes

- 📝 docs(README.md): fix typo in TODO list
- 🐛 fix(README.md): check off "Use config to set key" task

### Documentation

- 📝 docs(README.md): revise TODO list
- 📝 docs(README.md): add help.gif to Usage section
- 📝 docs(README.md): update file to reflect correct configuration file name.
- 📝 docs(config.yml): add configuration for welcome, new-issue-welcome, new-pr-welcome, and first-pr-merge behavior bots

### Features

- 📝 docs(CHANGELOG.md): add new features and changes in version 0.1.4
- ✨ feat(create_gif.sh): add script to create gif from asciinema recording
- ✨ feat(chat_config.py): add ability to set OpenAI API keys in config file using command line arguments

### Miscellaneous Tasks

- 🔍 chore(.gitignore): update .gitignore to exclude chatre.toml instead of apikey.toml
- 🔍 chore(.gitignore): add \*.cast to ignore list
- 🔧 chore(cliff.toml): add configuration file for git-cliff
- 🚀 chore(README.md): add 'Use config to set key' to the to-do list
- 🔖 chore(pyproject.toml): bump up version to 0.1.6

### Styling

- 🎨 style(tutorial/help.gif): add help.gif to tutorial directory

src/chat_research/reader.py
import base64
import datetime
import re
from pathlib import Path

import openai
import requests
import tenacity
import tiktoken
from loguru import logger

from .export import export
from .utils import load_config


class BaseReader:
    def __init__(self, root_path, language, file_format, save_image):
        self.root_path = Path(root_path)
        self.language = language
        self.file_format = file_format

        self.config, self.chat_api_list = load_config()
        self.cur_api = 0

        self.gitee_key = self.config["Gitee"]["api"] if save_image else ""

        self.max_token_num = 4096
        self.encoding = tiktoken.get_encoding("gpt2")
        self.token_usage = 0

    def summary_with_chat(self, paper_list, key_words):
        htmls = []
        for paper_index, paper in enumerate(paper_list):
            # 第一步先用title，abs，和introduction进行总结。
            text = ""
            text += "Title:" + paper.title
            text += "Url:" + paper.url
            text += "Abstrat:" + paper.abs
            text += "Paper_info:" + paper.section_text_dict["paper_info"]
            # intro
            text += list(paper.section_text_dict.values())[0]
            chat_summary_text = ""

            try:
                chat_summary_text = self.chat_summary(text=text, key_words=key_words)
            except Exception as e:
                logger.warning(f"summary_error: {e}")
                if "maximum context" in str(e):
                    current_tokens_index = (
                        str(e).find("your messages resulted in")
                        + len("your messages resulted in")
                        + 1
                    )
                    offset = int(
                        str(e)[current_tokens_index : current_tokens_index + 4]
                    )
                    summary_prompt_token = offset + 1000 + 150
                    chat_summary_text = self.chat_summary(
                        text=text,
                        key_words=key_words,
                        summary_prompt_token=summary_prompt_token,
                    )
                else:
                    raise e

            htmls.append("## Paper:" + str(paper_index + 1))
            htmls.append("\n\n\n")
            htmls.append(chat_summary_text)

            # 第二步总结方法：
            # TODO，由于有些文章的方法章节名是算法名，所以简单的通过关键词来筛选，很难获取，后面需要用其他的方案去优化。
            method_key = ""
            for parse_key in paper.section_text_dict.keys():
                if "method" in parse_key.lower() or "approach" in parse_key.lower():
                    method_key = parse_key
                    break

            if method_key != "":
                text = ""
                method_text = ""
                summary_text = ""
                summary_text += "<summary>" + chat_summary_text
                # methods
                method_text += paper.section_text_dict[method_key]
                text = summary_text + "\n\n<Methods>:\n\n" + method_text
                chat_method_text = ""
                try:
                    chat_method_text = self.chat_method(text=text, key_words=key_words)
                except Exception as e:
                    logger.error(f"method_error: {e}")
                    if "maximum context" in str(e):
                        current_tokens_index = (
                            str(e).find("your messages resulted in")
                            + len("your messages resulted in")
                            + 1
                        )
                        offset = int(
                            str(e)[current_tokens_index : current_tokens_index + 4]
                        )
                        method_prompt_token = offset + 800 + 150
                        chat_method_text = self.chat_method(
                            text=text,
                            key_words=key_words,
                            method_prompt_token=method_prompt_token,
                        )
                htmls.append(chat_method_text)
            else:
                chat_method_text = ""
            htmls.append("\n" * 4)

            # 第三步总结全文，并打分：
            conclusion_key = ""
            for parse_key in paper.section_text_dict.keys():
                if "conclu" in parse_key.lower():
                    conclusion_key = parse_key
                    break

            text = ""
            conclusion_text = ""
            summary_text = ""
            summary_text += (
                "<summary>"
                + chat_summary_text
                + "\n <Method summary>:\n"
                + chat_method_text
            )
            if conclusion_key != "":
                # conclusion
                conclusion_text += paper.section_text_dict[conclusion_key]
                text = summary_text + "\n\n<Conclusion>:\n\n" + conclusion_text
            else:
                text = summary_text

            chat_conclusion_text = ""
            try:
                chat_conclusion_text = self.chat_conclusion(
                    text=text, key_words=key_words
                )
            except Exception as e:
                logger.info(f"conclusion_error: {e}")
                if "maximum context" in str(e):
                    current_tokens_index = (
                        str(e).find("your messages resulted in")
                        + len("your messages resulted in")
                        + 1
                    )
                    offset = int(
                        str(e)[current_tokens_index : current_tokens_index + 4]
                    )
                    conclusion_prompt_token = offset + 800 + 150
                    chat_conclusion_text = self.chat_conclusion(
                        text=text,
                        key_words=key_words,
                        conclusion_prompt_token=conclusion_prompt_token,
                    )
            htmls.append(chat_conclusion_text)
            htmls.append("\n" * 4)

            # # 整合成一个文件，打包保存下来。
            date_str = str(datetime.datetime.now())[:13].replace(" ", "-")
            export_path = self.root_path / "export"

            if not export_path.exists():
                export_path.mkdir(parents=True, exist_ok=True)

            file_name = (
                Path(export_path) / f"{date_str}-{self.validateTitle(paper.title[:80])}"
            )

            export(
                content="\n".join([item.strip() for item in htmls]),
                file_name=file_name.with_suffix(f".{self.file_format}"),
            )

            htmls = []

    @tenacity.retry(
        wait=tenacity.wait_exponential(multiplier=1, min=4, max=10),
        stop=tenacity.stop_after_attempt(5),
        reraise=True,
    )
    def chat_conclusion(self, text, key_words, conclusion_prompt_token=800):
        openai.api_key = self.chat_api_list[self.cur_api]
        self.cur_api += 1
        self.cur_api = (
            0 if self.cur_api >= len(self.chat_api_list) - 1 else self.cur_api
        )
        text_token = len(self.encoding.encode(text))
        clip_text_index = int(
            len(text) * (self.max_token_num - conclusion_prompt_token) / text_token
        )
        clip_text = text[:clip_text_index]

        messages = [
            {
                "role": "system",
                "content": f"You are a reviewer in the field of [{key_words}] and you need to critically review this article",
            },
            # chatgpt 角色
            {
                "role": "assistant",
                "content": "This is the <summary> and <conclusion> part of an English literature, where <summary> you have already summarized, but <conclusion> part, I need your help to summarize the following questions:"
                + clip_text,
            },
            # 背景知识，可以参考OpenReview的审稿流程
            {
                "role": "user",
                "content": """
                 8. Make the following summary.Be sure to use {} answers (proper nouns need to be marked in English).
                    - (1):What is the significance of this piece of work?
                    - (2):Summarize the strengths and weaknesses of this article in three dimensions: innovation point, performance, and workload.
                    .......
                 Follow the format of the output later:
                 8. Conclusion: \n\n
                    - (1):xxx;\n
                    - (2):Innovation point: xxx; Performance: xxx; Workload: xxx;\n

                 Be sure to use {} answers (proper nouns need to be marked in English), statements as concise and academic as possible, do not repeat the content of the previous <summary>, the value of the use of the original numbers, be sure to strictly follow the format, the corresponding content output to xxx, in accordance with \n line feed, ....... means fill in according to the actual requirements, if not, you can not write.
                 """.format(
                    self.language, self.language
                ),
            },
        ]
        response = openai.ChatCompletion.create(
            model="gpt-3.5-turbo",
            # prompt需要用英语替换，少占用token。
            messages=messages,
        )
        result = ""
        for choice in response.choices:
            result += choice.message.content

        result = self.format_text(result)
        logger.trace(f"conclusion_result:\n{result}")
        self.report_token_usage(response)

        return result

    @tenacity.retry(
        wait=tenacity.wait_exponential(multiplier=1, min=4, max=10),
        stop=tenacity.stop_after_attempt(5),
        reraise=True,
    )
    def chat_method(self, text, key_words, method_prompt_token=800):
        openai.api_key = self.chat_api_list[self.cur_api]
        self.cur_api += 1
        self.cur_api = (
            0 if self.cur_api >= len(self.chat_api_list) - 1 else self.cur_api
        )
        text_token = len(self.encoding.encode(text))
        clip_text_index = int(
            len(text) * (self.max_token_num - method_prompt_token) / text_token
        )
        clip_text = text[:clip_text_index]

        messages = [
            {
                "role": "system",
                "content": f"You are a researcher in the field of [{key_words}] who is good at summarizing papers using concise statements",
            },
            # chatgpt 角色
            {
                "role": "assistant",
                "content": "This is the <summary> and <Method> part of an English document, where <summary> you have summarized, but the <Methods> part, I need your help to read and summarize the following questions."
                + clip_text,
            },
            # 背景知识
            {
                "role": "user",
                "content": """
                 7. Describe in detail the methodological idea of this article. Be sure to use {} answers (proper nouns need to be marked in English). For example, its steps are.
                    - (1):...
                    - (2):...
                    - (3):...
                    - .......
                 Follow the format of the output that follows:
                 7. Methods: \n\n
                    - (1):xxx;\n
                    - (2):xxx;\n
                    - (3):xxx;\n
                    ....... \n\n

                 Be sure to use {} answers (proper nouns need to be marked in English), statements as concise and academic as possible, do not repeat the content of the previous <summary>, the value of the use of the original numbers, be sure to strictly follow the format, the corresponding content output to xxx, in accordance with \n line feed, ....... means fill in according to the actual requirements, if not, you can not write.
                 """.format(
                    self.language, self.language
                ),
            },
        ]
        response = openai.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages=messages,
        )
        result = ""
        for choice in response.choices:
            result += choice.message.content

        result = self.format_text(result)
        logger.trace(f"method_result:\n{result}")
        self.report_token_usage(response)

        return result

    @tenacity.retry(
        wait=tenacity.wait_exponential(multiplier=1, min=4, max=10),
        stop=tenacity.stop_after_attempt(5),
        reraise=True,
    )
    def chat_summary(self, text, key_words, summary_prompt_token=1100):
        openai.api_key = self.chat_api_list[self.cur_api]
        self.cur_api += 1
        self.cur_api = (
            0 if self.cur_api >= len(self.chat_api_list) - 1 else self.cur_api
        )
        text_token = len(self.encoding.encode(text))
        clip_text_index = int(
            len(text) * (self.max_token_num - summary_prompt_token) / text_token
        )
        clip_text = text[:clip_text_index]

        messages = [
            {
                "role": "system",
                "content": f"You are a researcher in the field of [{key_words}]  who is good at summarizing papers using concise statements",
            },
            {
                "role": "assistant",
                "content": "This is the title, author, link, abstract and introduction of an English document. I need your help to read and summarize the following questions: "
                + clip_text,
            },
            {
                "role": "user",
                "content": f"""
                 1. Mark the title of the paper
                 2. list all the authors' names (use English)
                 3. mark the first author's affiliation (use English)
                 4. mark the keywords of this article (use English)
                 5. link to the paper, Github code link (if available, fill in Github:None if not)
                 6. summarize according to the following four points.Be sure to use {self.language} answers (proper nouns need to be marked in English)
                    - (1):What is the research background of this article?
                    - (2):What are the past methods? What are the problems with them? Is the approach well motivated?
                    - (3):What is the research methodology proposed in this paper?
                    - (4):On what task and what performance is achieved by the methods in this paper? Can the performance support their goals?
                 Follow the format of the output that follows:
                 1. Title: xxx\n\n
                 2. Authors: xxx\n\n
                 3. Affiliation: xxx\n\n
                 4. Keywords: xxx\n\n
                 5. Urls: xxx or xxx , xxx \n\n
                 6. Summary: \n\n
                    - (1):xxx;\n
                    - (2):xxx;\n
                    - (3):xxx;\n
                    - (4):xxx.\n\n

                 Be sure to use {self.language} answers (proper nouns need to be marked in English), statements as concise and academic as possible, do not have too much repetitive information, numerical values using the original numbers, be sure to strictly follow the format,
                 the corresponding content output to xxx, in accordance with \n line feed.
                 """,
            },
        ]

        response = openai.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages=messages,
        )
        result = ""
        for choice in response.choices:
            result += choice.message.content

        result = self.format_text(result)
        logger.trace(f"summary_result:\n{result}")

        self.report_token_usage(response)

        return result

    def validateTitle(self, title):
        # 将论文的乱七八糟的路径格式修正
        rstr = r"[\/\\\:\*\?\"\<\>\|]"  # '/ \ : * ? " < > |'
        new_title = re.sub(rstr, "_", title)  # 替换为下划线
        return new_title

    @staticmethod
    def format_text(text):
        result = ""
        for line in text.split("\n"):
            result += line.strip() + "\n"
        return result

    def show_token_usage(self):
        money = self.token_usage / 1000 * 0.002
        logger.info(f"total_token_used: {self.token_usage} ({money} USD)")

    def report_token_usage(self, response):
        logger.info(f"prompt_token_used: {response.usage.prompt_tokens}")
        logger.info(f"completion_token_used: {response.usage.completion_tokens}")
        logger.info(f"total_token_used: {response.usage.total_tokens}")
        logger.info(f"response_time: { response.response_ms / 1000.0}s")
        self.token_usage += response.usage.total_tokens

    @tenacity.retry(
        wait=tenacity.wait_exponential(multiplier=1, min=4, max=10),
        stop=tenacity.stop_after_attempt(5),
        reraise=True,
    )
    def upload_gitee(self, image_path, image_name="", ext="png"):
        """
        上传到码云
        :return:
        """
        with open(image_path, "rb") as f:
            base64_data = base64.b64encode(f.read())
            base64_content = base64_data.decode()

        date_str = (
            str(datetime.datetime.now())[:19].replace(":", "-").replace(" ", "-")
            + "."
            + ext
        )
        path = image_name + "-" + date_str

        payload = {
            "access_token": self.gitee_key,
            "owner": self.config["Gitee"]["owner"],
            "repo": self.config["Gitee"]["repo"],
            "path": self.config["Gitee"]["path"],
            "content": base64_content,
            "message": "upload image",
        }
        # 这里需要修改成你的gitee的账户和仓库名，以及文件夹的名字：
        url = (
            "https://gitee.com/api/v5/repos/"
            + self.config["Gitee"]["owner"]
            + "/"
            + self.config["Gitee"]["repo"]
            + "/contents/"
            + self.config["Gitee"]["path"]
            + "/"
            + path
        )
        rep = requests.post(url, json=payload).json()
        logger.info(f"{rep=}")
        if "content" in rep.keys():
            image_url = rep["content"]["download_url"]
        else:
            image_url = (
                r"https://gitee.com/api/v5/repos/"
                + self.config["Gitee"]["owner"]
                + "/contents/"
                + self.config["Gitee"]["path"]
                + "/"
                + path
            )

        return image_url
